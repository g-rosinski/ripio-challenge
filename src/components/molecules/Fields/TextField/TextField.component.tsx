import React, { InputHTMLAttributes, ChangeEvent } from "react"
import useInputHelper from "../../../../hooks/useInputHelper/useInputHelper.hook"
import { InputGroup, Label, HelperText, Input } from "../../../atoms"

type TextFieldProps = Omit<InputHTMLAttributes<HTMLInputElement>, 'type' | 'onChange'> & {
    value: string,
    onChange: (e: ChangeEvent<HTMLInputElement>) => void,
    label?: string,
    helper?: string,
    error?: string,
}

const TextField: React.FC<TextFieldProps> = ({
    value, 
    label, 
    helper = "", 
    error,
    disabled,
    onChange,
    name,
    ...props
}) => {
    const { displayHelperText, helperText, helperStatus} = useInputHelper({
        disabled,
        error,
        info: helper
    })
    
    return (
        <InputGroup>
            {label && <Label htmlFor={name}>{label}</Label>}
            <Input 
                id={name}
                type={'text'}
                name={name}
                value={value}
                onChange={onChange}
                disabled={disabled}
                {...props}
            />
            { displayHelperText && <HelperText status={helperStatus}>{ helperText }</HelperText> }
        </InputGroup>
    )
}

export default TextField