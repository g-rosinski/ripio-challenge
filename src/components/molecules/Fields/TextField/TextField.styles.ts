import styled from "styled-components"

export const Input = styled.input`
    font-size: 14px;
    padding: 3px 5px;
    border: none;
    border-bottom: 1px solid;
    min-width: 100px;
`