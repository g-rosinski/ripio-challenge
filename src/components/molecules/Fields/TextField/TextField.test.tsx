import { render, screen, waitFor } from '@testing-library/react'
import userEvent from '@testing-library/user-event'
import React from 'react'
import TextField from './TextField.component'

describe('<TextField/>', () => {

    const onChangeMock = jest.fn()

    beforeEach(() => {
        jest.clearAllMocks()
    })

    test('Se espera que se reenderice correctamente', () => {
        const {container} = render(<TextField value={""} onChange={onChangeMock} />)
        
        expect(screen.getByRole('textbox')).toHaveValue("")
        expect(screen.getByRole('textbox')).toHaveProperty("type",'text')
        expect(container).toMatchSnapshot()
    })

    test('Se espera que prop onChange al tipear un valor', async () => {
        render(<TextField value={""} onChange={onChangeMock} />)
        
        const input = screen.getByRole('textbox')
        expect(input).toHaveValue("")
        
        userEvent.type(input, 'Lionel')
        expect(onChangeMock).toHaveBeenCalledTimes(6)
    })

    
    test('Se espera que si disabled es true, la prop onChange no sea llamada', async () => {
        render(<TextField value={""} onChange={onChangeMock} disabled />)

        const input = screen.getByRole('textbox')
        
        userEvent.type(input, 'Lionel')
        expect(onChangeMock).not.toBeCalled()
    })

    test('Se espera que prop value se cambie al setear un value nuevo', async () => {
        const {rerender} = render(<TextField value={""} onChange={onChangeMock} />)
        
        const input = screen.getByRole('textbox')
        expect(input).toHaveValue("")
        
        rerender(<TextField value={"Lionel"} onChange={onChangeMock} />)
        expect(input).toHaveValue('Lionel')
    })

    test('Se espera que al setear un label se reenderice un label', async () => {
        render(<TextField label='Nombre' value={""} onChange={onChangeMock} />)
        
        expect(screen.getByText('Nombre')).toBeInTheDocument()
    })

    test('Se espera que al setear un helper se reenderice el helperText', async () => {
        render(<TextField helper='Ingrese un nombre' value={""} onChange={onChangeMock} />)
        
        expect(screen.getByText('Ingrese un nombre')).toBeInTheDocument()
    })

    test('Se espera que al setear un error, helper se reenderice el error en el helperText', async () => {
        render(<TextField error='Nombre no valido' helper='Ingrese un nombre' value={""} onChange={onChangeMock} />)
        
        expect(screen.getByText('Nombre no valido')).toBeInTheDocument()
        expect(screen.queryByText('Ingrese un nombre')).not.toBeInTheDocument()
    })

})