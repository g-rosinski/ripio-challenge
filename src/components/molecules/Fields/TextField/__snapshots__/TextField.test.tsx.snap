// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TextField/> Se espera que se reenderice correctamente 1`] = `
<div>
  <div
    class="sc-gswNZR fGuKlY"
  >
    <input
      class="sc-gKPRtg crDhfD"
      type="text"
      value=""
    />
  </div>
</div>
`;
